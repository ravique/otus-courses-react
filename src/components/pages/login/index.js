import React from "react";
import SmartLoginContainer from "./login_form";
import Content from "../../container/content";

const Login = () => (
    <Content>
        <h2>Login:</h2>
        <SmartLoginContainer/>
    </Content>
);

export default Login;