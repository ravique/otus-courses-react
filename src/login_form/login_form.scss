.login_form {
  width: 500px;
  margin: 0 auto;

  &__input {
    width: 100%;
  }

  &__label {
    display: block;
    margin: 10px 0 0 0;
  }

  &__errors {
    color: $red;

    &__message {
      display: block;
    }
  }

}

