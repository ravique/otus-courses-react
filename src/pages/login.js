import React from "react";
import LoginContainer from "../login_form/login_form";
import Content from "../container/content";

const Login = () => (
    <Content>
        <h2>Login:</h2>
        <LoginContainer/>
    </Content>
);

export default Login