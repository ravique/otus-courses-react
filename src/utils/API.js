import axios from "axios";

axios.defaults.xsrfCookieName = 'csrftoken';
axios.defaults.xsrfHeaderName = 'X-CSRFToken';

console.log(process.env.REACT_APP_API_URL);

export default axios.create({
  baseURL: process.env.REACT_APP_API_URL,
});