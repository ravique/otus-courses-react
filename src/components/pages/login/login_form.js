import React, {Component} from "react";
import {Redirect} from "react-router-dom";

import API from "../../../utils/API";
import ThrowErrors from "../../throw_errors/index";
import {FormInput, FormLabel} from "../../form/index";

export default class LoginContainer extends Component {
    constructor() {
        super();

        this.state = {
            loggedIn: false,
            errors: '',
            username: '',
            password: ''
        };
    }

    handleChange = ({target: {name, value}}) => {
        this.setState({[name]: value});
    };

    handleSubmit = event => {
        event.preventDefault();
        API.post('login/', {"username": this.state.username, "password": this.state.password})
            .then(response => {
                    this.setState({
                        loggedIn: true
                    })
                }
            ).catch(
            errors => {
                if (errors.response) {
                    console.log(errors.response.data['errors']);
                    this.setState(
                        {
                            errors: errors.response.data['errors']
                        }
                    );
                }
            }
        )
    };

    render() {
        if (!this.state.loggedIn) {
            return (
                <form className="form" onSubmit={this.handleSubmit.bind(this)}>
                    <FormLabel name="login"/>
                    <FormInput
                        type="text"
                        name="username"
                        eventHandler={this.handleChange}
                        placeholder="Enter your username."
                        value={this.state.username}
                        error={this.state.errors.username}
                    />
                    <ThrowErrors message={this.state.errors.username}/>

                    <FormLabel name="password"/>
                    <FormInput
                        type="password"
                        name="password"
                        eventHandler={this.handleChange}
                        placeholder="Enter your password."
                        value={this.state.password}
                        error={this.state.errors.password}
                    />
                    <ThrowErrors message={this.state.errors.password}/>

                    <ThrowErrors message={this.state.errors.non_field_errors}/>

                    <button type="submit" className="button button_full_width">Login</button>
                </form>
            )
        } else {
            return (
                <Redirect to="/"/>
            )
        }
    }
}

