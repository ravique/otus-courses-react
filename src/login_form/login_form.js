import React, {Component} from "react";
import {Redirect} from "react-router-dom";

import API from "../utils/API";
import ThrowErrors from "../throw_errors/throw_errors";
import {FormLabel} from "../form/form";

export default class LoginContainer extends Component {
    constructor(props) {
        super(props);

        this.state = {
            loggedIn: false,
            errors: '',
            username: '',
            password: ''
        };
    }

    handleChange(event) {
        this.setState({
            [event.target.name]: event.target.value
        });
    }

    handleSubmit = event => {
        event.preventDefault();
        API.post('login/', {"username": this.state.username, "password": this.state.password})
            .then(response => {
                    this.setState({
                        loggedIn: true
                    })
                }
            ).catch(
            errors => {
                if (errors.response) {
                    console.log(errors.response.data['errors']);
                    this.setState(
                        {
                            errors: errors.response.data['errors']
                        }
                    );
                }
            }
        )
    };

    render() {

        if (!this.state.loggedIn) {
            return (
                <div className="form" onSubmit={this.handleSubmit.bind(this)}>
                    <form className="Login">
                        <FormLabel name="login"/>
                        <input
                            type="text"
                            className="form__input"
                            name="username"
                            onChange={this.handleChange.bind(this)}
                            placeholder="Enter your username."
                            value={this.state.username}/>
                        <ThrowErrors message={this.state.errors.username}/>

                        <FormLabel name="password"/>
                        <input
                            type="password"
                            className="form__input"
                            name="password"
                            onChange={this.handleChange.bind(this)}
                            value={this.state.password}
                            placeholder="Enter your password."/>
                        <ThrowErrors message={this.state.errors.password}/>

                        <ThrowErrors message={this.state.errors.non_field_errors}/>

                        <button type="submit" className="button button_full_width">Login</button>
                    </form>
                </div>

            )
        } else {
            return (
                <Redirect to="/"/>
            )
        }
    }
}

