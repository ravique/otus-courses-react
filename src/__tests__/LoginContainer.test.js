import React from 'react';
import {mount} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import '../../jest.config'
import {MemoryRouter} from "react-router-dom";

import {LoginContainer} from '../components/pages/login/login_form'
import API from "../utils/API";
import MockAdapter from 'axios-mock-adapter';
import {fakeEvent} from "../__mocks__";



describe('LoginContainer', () => {

    let output;
    
    beforeEach(() => {
        output = mount(
            <MemoryRouter>
                <LoginContainer/>
            </MemoryRouter>
        ).find('LoginContainer');
    });

    it('should render login container', () => {
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('should login', () => {
        expect(output.state().loggedIn).toBeFalsy();
    });

    it('should change state on fields change', () => {
            const mockChangeUsername = {
                target: {
                    name: 'username',
                    value: 'John'
                }
            };

            const mockChangePassword = {
                target: {
                    name: 'password',
                    value: '32768'
                }
            };

            const expected = {
                username: 'John',
                password: '32768',
                loggedIn: false
            };

            output.instance().handleChange(mockChangeUsername);
            output.instance().handleChange(mockChangePassword);

            expect(output.state()).toMatchObject(expected);
        }
    );

    it('should login', async () => {

        let mock = new MockAdapter(API);
        mock.onPost('login/').reply(200);

        await output.instance().handleSubmit(fakeEvent);

        expect(output.state().loggedIn).toEqual(true);
    });

    it('should fail login', async () => {

        let mock2 = new MockAdapter(API);
        mock2.onPost('login/').reply(400, {"errors":{"non_field_errors":["Incorrect email or password."]}});

        await output.instance().handleSubmit(fakeEvent);

        expect(output.state().loggedIn).toEqual(false);
        expect(output.state().errors.non_field_errors).toContain('Incorrect email or password.');
        expect(output.text()).toContain('Incorrect email or password.');
    })
});