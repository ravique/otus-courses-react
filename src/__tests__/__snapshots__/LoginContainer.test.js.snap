// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer should render login container 1`] = `
<LoginContainer>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormLabel
      name="login"
    >
      <label
        className="form__label"
        htmlFor="login"
      >
        Login
      </label>
    </FormLabel>
    <FormInput
      eventHandler={[Function]}
      name="username"
      placeholder="Enter your username."
      type="text"
      value=""
    >
      <input
        className="form__input"
        name="username"
        onChange={[Function]}
        placeholder="Enter your username."
        type="text"
        value=""
      />
    </FormInput>
    <ThrowErrors />
    <FormLabel
      name="password"
    >
      <label
        className="form__label"
        htmlFor="password"
      >
        Password
      </label>
    </FormLabel>
    <FormInput
      eventHandler={[Function]}
      name="password"
      placeholder="Enter your password."
      type="password"
      value=""
    >
      <input
        className="form__input"
        name="password"
        onChange={[Function]}
        placeholder="Enter your password."
        type="password"
        value=""
      />
    </FormInput>
    <ThrowErrors />
    <ThrowErrors />
    <button
      className="button button_full_width"
      type="submit"
    >
      Login
    </button>
  </form>
</LoginContainer>
`;
